lobbyboy/contrib/provider/digitalocean.py

```diff
@@ -1,227 +1,132 @@
-        action1 = actions[0]
-        check = 0
-        while 1:
-            if check <= 2:
-                time.sleep(5)
-            else:
-                time.sleep(1)
-
-            logger.debug("{} time check for action: {}".format(check, action1))
-            check += 1
-            action1.load()
-            logger.debug("check result: {}".format(action1))
-            # Once it shows "completed", droplet is up and running
-            if action1.status == "completed":
-                break
-
-        droplet.load()
-        event.set()
-        server_info_path = str(server_data / "server.json")
-        self._dump_info(droplet, server_info_path)
-        chan.send(
-            "New server {}({}) created!\r\n".format(
-                server_name, droplet.ip_address
-            ).encode()
-        )
-        # wait for server to start up...
-        chan.send("Waiting server to boot...\r\n".encode())
-        self.block_test_server_connectable(droplet.ip_address, 22)
-        return server_name, droplet.ip_address
-
-    def block_test_server_connectable(self, ip, port):
-
-        location = (ip, int(port))
-        result = 1
-        count = 1
-        while result != 0:
-            a_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            a_socket.settimeout(2)
-            result = a_socket.connect_ex(location)
-            a_socket.close()
-            logger.debug("{} check, reuslt is {}".format(count, result))
-            time.sleep(1)
-            count += 1
-
-    def _ask_user_custmize_server(self, chan):
-        manually_create_choice = "Manually choose a new droplet to create..."
-        favorites = [manually_create_choice] + self.provider_config["favorite_droplets"]
-        choosed, _ = self.choose_option(
-            "Please choose new droplet to create: ", favorites, chan
+        self.time_process_action(channel, self.droplet_is_up, uninitialized_droplet=droplet)
+        return self.prepare_after_server_created(channel, droplet, server_workspace, server_name)
+
+    def prepare_after_server_created(self, c: Channel, d: Droplet, workspace: Path, server_name: str) -> LBServerMeta:
+        # todo: support use startup script after server created
+        # load full information from digitalocean before use this droplet to other operations
+        d.load()
+        send_to_channel(c, f"New server {server_name} (IP: {d.ip_address}) created!")
+
+        # save server info to local first
+        droplet_meta = lb_dict_factory(d.__dict__, ignore_fields=["tokens"], ignore_rules=lambda x: x.startswith("_"))
+        self.save_raw_server(droplet_meta, workspace)
+
+        # wait for server to startup(check port is alive or not)
+        send_to_channel(c, "Waiting for server to boot...")
+        self.time_process_action(c, port_is_open, ip=d.ip_address)
+        send_to_channel(c, f"Server {server_name} has boot successfully!")
+
+        return LBServerMeta(
```

Comment:
better to return native Python type instead of our defined type, so that user don't have to know the internal class of lobbyboy, to make the new provider implementation easier.

Reply:
lobbyboy needs to manage many different kinds of servers, so maybe it's better for user that do something according to the rules of lobbyboy. users know in the process of implementing their own provider, at the end of `create_server`, they need to give lobbyboy a fixed thing that lobbyboy known instead of python type, python type is too open, such as the order of tuple/list, key of dict, they are unstable to make mistake easily in some cases.